### PR TITLE
Add update code path for repository tags

### PR DESCRIPTION
**Pull request by A-Hilaly**
Issue https://github.com/aws-controllers-k8s/community/issues/1137

Description of changes:
Add update code path for ECR repositories
- Regenerate controller with the code-gen latest commit
- Add custom hooks for tags comparison and update code path
- Add e2e tests for repository tags update

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: config/rbac/service-account.yaml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
```
Comment (RedbackThomson) [inline config/rbac/service-account.yaml:1]: Would you mind generating this new service account stuff in a separate PR, just to clean this one up a little bit?

File: pkg/resource/repository/custom_update_api.go [hunk 2]
```diff
@@ -193,3 +199,55 @@ func (rm *resourceManager) deleteLifecyclePolicy(
...
+	desired *resource,
+) (err error) {
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.syncRepositoryTags")
+	defer exit(err)
+
+	added, updated, removed := computeTagsDelta(latest.ko.Spec.Tags, desired.ko.Spec.Tags)
+
+	// There are no API calls to update an existing tag. To update a tag we will have to first
+	// delete it and then recreate it with the new value.
```
Comment (RedbackThomson) [inline pkg/resource/repository/custom_update_api.go:216]: If you call `TagResource` with the key of an existing tag, will that not update it? I see that behaviour is not documented in the [docs](https://docs.aws.amazon.com/sdk-for-go/api/service/ecr/#ECR.TagResource)

File: test/e2e/tests/test_repository.py [hunk 3]
```diff
@@ -174,6 +184,63 @@ def test_repository_lifecycle_policy(self, ecr_client):
...
+        # Check ECR repository exists
+        exists = self.repository_exists(ecr_client, resource_name)
+        assert exists
+
+        # Update CR
+        tags = [{
+            "key": "k1",
+            "value": "v1",
+        }]
+        cr["spec"]["tags"] = tags
```
Comment (RedbackThomson) [inline test/e2e/tests/test_repository.py:227]: I think we could also benefit from another step that then updates this tag, and then another which deletes it - just to check that all code paths are working properly.

**Review by RedbackThomson — CHANGES_REQUESTED** [review]
Nice. Really smooth implementation.

File: config/rbac/service-account.yaml (hunk 1 above)
Comment (A-Hilaly) [reply config/rbac/service-account.yaml:1]: Regenerated controller in https://github.com/aws-controllers-k8s/ecr-controller/pull/32

**Comment by A-Hilaly** [comment]
waiting for #32 merge /hold

**Comment by A-Hilaly** [comment]
waiting for #32 merge
/hold

File: pkg/resource/repository/custom_update_api.go (hunk 2 above)
Comment (A-Hilaly) [reply pkg/resource/repository/custom_update_api.go]: Good call! I just tested `TagResource` and apparently it updates the tags aswell! :+1:

File: pkg/resource/repository/hook.go [hunk 4]
```diff
@@ -29,14 +33,108 @@ func (rm *resourceManager) setResourceAdditionalFields(
...
+	if listTagsForResourceResponse.Tags != nil {
+		tags := make([]*v1alpha1.Tag, 0, len(listTagsForResourceResponse.Tags))
+		for _, tag := range listTagsForResourceResponse.Tags {
+			tags = append(tags, &v1alpha1.Tag{
+				Key:   tag.Key,
+				Value: tag.Value,
+			})
+		}
+		ko.Spec.Tags = tags
+	}
```
Comment (jaypipes) [inline pkg/resource/repository/hook.go:65]: Two things.. this isn't handling pagination. Yes, I know the default number of returned tags is 100 and it's unlikely anyone is going to use more than 100 tags on a repository, but you never know (and that bug would be quite subtle and hard to debug if someone were to run into it).

Secondly, consider creating a separate helper function that calls `ListTagsForResource` and returns a `[]*v1alpha1.Tag` structs. Then here, you would do something like:

```go
ko.Spec.Tags = getRepositoryTags((*string)(ko.Status.ACKResourceMetadata.ARN))
```

**Review by jaypipes — CHANGES_REQUESTED** [review]

File: pkg/resource/repository/hook.go (hunk 4 above)
Comment (A-Hilaly) [reply pkg/resource/repository/hook.go:65]: > This isn't handling pagination. Yes, I know the default number of returned tags is 100 and it's unlikely anyone is going to use more than 100 tags on a repository, but you never know (and that bug would be quite subtle and hard to debug if someone were to run into it).

Unfortunately `ListTagsForResource` call doesn't support pagination...

> Secondly, consider creating a separate helper function that calls ListTagsForResource and returns a []*v1alpha1.Tag structs. Then here, you would do something like:

++

File: pkg/resource/repository/custom_update_api.go [hunk 5]
```diff
@@ -193,3 +199,48 @@ func (rm *resourceManager) deleteLifecyclePolicy(
...
+func (rm *resourceManager) syncRepositoryTags(
+	ctx context.Context,
+	latest *resource,
+	desired *resource,
+) (err error) {
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.syncRepositoryTags")
+	defer exit(err)
+
+	added, updated, removed := computeTagsDelta(latest.ko.Spec.Tags, desired.ko.Spec.Tags)
```
[the remark this answers is not in the document]
Comment (A-Hilaly) [reply pkg/resource/repository/custom_update_api.go:213]: Added e2e test for updated tags

File: pkg/resource/repository/hook.go (hunk 4 above)
Comment (jaypipes) [reply pkg/resource/repository/hook.go]: > > This isn't handling pagination. Yes, I know the default number of returned tags is 100 and it's unlikely anyone is going to use more than 100 tags on a repository, but you never know (and that bug would be quite subtle and hard to debug if someone were to run into it).
> 
> Unfortunately `ListTagsForResource` call doesn't support pagination... I believe this API returns all the tags at ones and have probably limited the number of tags you can attach to a resource

**sigh**. Yet another AWS API inconsistency :) OK, thanks!

**Review by jaypipes — APPROVED** [review]
rock on, Amine. Nice work on this. ++

**Pull request merged**